Add collapse sidebar support

cpapazoglou opened the pull request:
#### Changes proposed in this Pull Request

* adds a collapse-sidebar component
* styles the sidebar collapsed state

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Calypso**
* run `yarn && yarn start`
* add `?flags=nav-unification`  to the URL
* check that you can:
1. expand / collapse the sidebar
2. when collapsed, click the site selector and the sidebar is expanded while site selector is shown
3. when collapsed, click a notice and the sidebar is expanded while you are navigated 
4. when collapsed, navigate to reader / me and back to My Sites

https://gifyu.com/image/8liS


**Wp-admin**
* apply D50420-code
* add the sticker to your site
* check your sandbox wp-admin
* compare the collapsed sidebar to the calypso one. It should be the same (except the notices / nudges missing)

* designs should be in par with https://www.figma.com/file/9cy0DIowQkPAu0rodb9jsL/WP-Admin-for-Dotcom?node-id=403%3A1356&viewport=-3368%2C411%2C1.8277816772460938

**Known Issues**
- [x] Menu Item Hover / Expanded state is not implemented
- [x] Clicking Site Card / Notices should expand the sidebar 
- [x] Site Card height should be 52px in calypso & wp-admin in all states


Fixes #45962

getdave commented on client/my-sites/sidebar-unified/index.jsx:
```diff
@@ -34,6 +36,13 @@ export const MySitesSidebarUnified = ( { path } ) => {
 	const menuItems = useSiteMenuItems();
 	const isAllDomainsView = useDomainsViewStatus();
 	const isRequestingMenu = useSelector( getIsRequestingAdminMenu );
+	const [ collapsed, setCollapsed ] = useState( isWithinBreakpoint( '<960px' ) ? true : false );
```
I wonder if you could set this based on a query param in the URL. Then you could sort of "share" the state of the sidebar from Calypso with WPAdmin.

cpapazoglou replied:
Great idea @getdave ! I think we can leave it for a next PR though?

obenland commented on client/my-sites/sidebar-unified/style.scss:
```diff
@@ -139,6 +144,78 @@ $font-size: rem( 14px );
+
+		.current-site__notices > a::before {
+			content: '\f534';
+			font-family: 'dashicons';
```
This doesn't seem to work for me on calypso.live. The icon doesn't show up and the banner is collapsed.

![image](https://user-images.githubusercontent.com/1398304/94714332-da29fc00-0300-11eb-8970-49d12a5dfe60.png)

cpapazoglou replied:
I am not sure how you get this.. It seems that the icon is not loaded? can you try in local just in case?

obenland commented:
I noticed that when toggling the sidebar, menu icons make quite a noticeable shift to the side, which I assume is because the collapsed menu is 8px wider than in wp-admin?

@sfougnier Do you think there's anything we can do to make that less noticeable? Is the wider width based on needing to accommodate the site card icon?

obenland commented:
A few more things I noticed:

- The Home menu item and the toggle are 52px high, while the rest of the menu items are 34px high.
- Submenus still get unfolded when a menu item is active. In wp-admin the collapsed menu stays pretty much the same, with the parent item marked as active if it or any of its submenus are active.
- It looks like wp-admin persists folding status through [a user setting](https://core.trac.wordpress.org/browser/tags/5.5.1/src/js/_enqueues/admin/common.js#L873) that [gets saved in a cookie](https://core.trac.wordpress.org/browser/tags/5.5.1/src/js/_enqueues/lib/cookies.js#L161). I'm not sure if there's a Calypso API that mirrors that, but if there is, that might be a way to persist the state between wp-admin & calypso as well as in between page loads & visits.
- The Site Switcher is currently unusable when the menu is collapsed. I'm not sure how to best deal with that. Maybe unfold when the site switcher was clicked and re-fold once a site was selected?
- The Media screen doesn't seem to have a max width, which results in images having a grey border with the unified nav, that becomes a thick gray border in the collapsed nav.

sfougnier commented:
>I noticed that when toggling the sidebar, menu icons make quite a noticeable shift to the side, which I assume is because the collapsed menu is 8px wider than in wp-admin?
>
>@sfougnier Do you think there's anything we can do to make that less noticeable? Is the wider width based on needing to accommodate the site card icon?

The wider width was to accommodate the site card icon, but I agree the jump is weird. This is how it looks when shrinking down the width by 8px. The vertical padding in the site card looks strange, but to avoid the vertical jumping we should just keep the height of all the nav items including the site card consistent with the expanded nav.

<img width="1351" alt="desktop - collapsed" src="https://user-images.githubusercontent.com/67021691/94736673-9bb13300-033a-11eb-8ad9-29268bb94478.png">

I have updated this in [Figma](https://www.figma.com/file/9cy0DIowQkPAu0rodb9jsL/WP-Admin-for-Dotcom?node-id=525%3A820)

> The Home menu item and the toggle are 52px high, while the rest of the menu items are 34px high.

Yes, also noticing that the size of the icon in the Switch Site area changes between the states.

> The Site Switcher is currently unusable when the menu is collapsed. I'm not sure how to best deal with that. Maybe unfold when the site switcher was clicked and re-fold once a site was selected?

Yes, I like this idea.

cpapazoglou commented:
Thanks @obenland for your review!

> * The Home menu item and the toggle are 52px high, while the rest of the menu items are 34px high.

@sfougnier figma shows 56px in expanded state and 52px in collapsed. Which of the two are we keeping? We will apply this height to `wp-admin` too. And please confirm that in mobile this height will be 72px.

> * Submenus still get unfolded when a menu item is active. In wp-admin the collapsed menu stays pretty much the same, with the parent item marked as active if it or any of its submenus are active.

I prevented submenus unfolding when sidebar is collapsed 08664ab. We will have to consolidate with https://github.com/Automattic/wp-calypso/pull/46046 regarding the flyout and colours.

> * It looks like wp-admin persists folding status through [a user setting]
(https://core.trac.wordpress.org/browser/tags/5.5.1/src/js/_enqueues/admin/common.js#L873) that [gets saved in a cookie](https://core.trac.wordpress.org/browser/tags/5.5.1/src/js/_enqueues/lib/cookies.js#L161). I'm not sure if there's a Calypso API that mirrors that, but if there is, that might be a way to persist the state between wp-admin & calypso as well as in between page loads & visits.

Indeed, I think we can ship this and check on an other PR how we can best implement this https://github.com/Automattic/wp-calypso/pull/46045#discussion_r498075486

> * The Site Switcher is currently unusable when the menu is collapsed. I'm not sure how to best deal with that. Maybe unfold when the site switcher was clicked and re-fold once a site was selected?

Probably that is a good idea. Let me work it out.

> * The Media screen doesn't seem to have a max width, which results in images having a grey border with the unified nav, that becomes a thick gray border in the collapsed nav.

Can you please share a screenshot for this

sfougnier commented:
>@sfougnier figma shows 56px in expanded state and 52px in collapsed. Which of the two are we keeping? We will apply this height to wp-admin too. And please confirm that in mobile this height will be 72px.

I looked into this - 52px was a result of a strange hidden layer that I've since fixed. For site card height, 56px should be the sizing in desktop and mobile should be 72px.

> Probably that is a good idea. Let me work it out. @sfougnier I guess we can do the same for notices / nudges?

Yes, sounds good.

cpapazoglou commented:
I have _stashed_ some changes needed regarding the Sidebar state here https://github.com/Automattic/wp-calypso/commit/41f883b1f7005909287bd54e43b89ef62797bd34

blackjackkent commented:
This looks great in testing on Calypso live. I think the only really blocking issue is the problem with the site switcher, as that seems like a negative user experience.

![image](https://user-images.githubusercontent.com/13437011/95381797-46cb6a80-08ae-11eb-9dbf-6c7b2bee9d9f.png)

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@cpapazoglou please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/928267). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

cpapazoglou commented:
I have made several changes:
1) fixed all UI inconsistencies both in calypso and wp-admin
2) added sidebar status to state
3) expand the sidebar if it is collapsed and user clicks in the sites/ notices area

mreishus commented on client/state/ui/test/selectors.js:
```diff
@@ -241,4 +242,55 @@ describe( 'selectors', () => {
+
+			expect( selected ).to.be.true;
+		} );
+	} );
```
`yarn run test` here gives me an error: `    ReferenceError: isSectionIsomorphic is not defined`.  Maybe this was a reference copy/paste left in?

cpapazoglou replied:
Thanks for pointing this out, it should be something with a wrong merge conflict resolution. This code is not mine nor it exists on master, I removed it.
Not sure why didn't I try tests in this branch.. Now `yarn jest -c=test/client/jest.config.js client/state/ui/test/selectors.js` passes

mreishus commented:
Suggestion: Reuse the existing name of the action creator
```diff
diff --git a/client/my-sites/current-site/index.jsx b/client/my-sites/current-site/index.jsx
index 8b84c6fea7..3984dabe16 100644
--- a/client/my-sites/current-site/index.jsx
+++ b/client/my-sites/current-site/index.jsx
@@ -37,12 +37,12 @@ class CurrentSite extends Component {
                anySiteSelected: PropTypes.array,
                forceAllSitesView: PropTypes.bool,
                sidebarIsCollapsed: PropTypes.bool,
-               expandTheSidebar: PropTypes.func.isRequired,
+               expandSidebar: PropTypes.func.isRequired,
        };
 
        switchSites = ( event ) => {
                if ( isEnabled( 'nav-unification' ) && this.props.sidebarIsCollapsed ) {
-                       this.props.expandTheSidebar();
+                       this.props.expandSidebar();
                }
                event.preventDefault();
                event.stopPropagation();
@@ -136,6 +136,6 @@ export default connect(
        {
                recordGoogleEvent,
                setLayoutFocus,
-               expandTheSidebar: () => expandSidebar(),
+               expandSidebar,
        }
 )( localize( CurrentSite ) );
```

mreishus commented:
The frontend is working properly for me in calypso locally, including the site switcher automatically triggering expansion when used in collapsed state. :heavy_check_mark:

cpapazoglou reviewed (commented):
> Suggestion: Reuse the existing name of the action creator
> 
> ```diff
> diff --git a/client/my-sites/current-site/index.jsx b/client/my-sites/current-site/index.jsx
> index 8b84c6fea7..3984dabe16 100644
> --- a/client/my-sites/current-site/index.jsx
> +++ b/client/my-sites/current-site/index.jsx
> @@ -37,12 +37,12 @@ class CurrentSite extends Component {
>                 anySiteSelected: PropTypes.array,
>                 forceAllSitesView: PropTypes.bool,
>                 sidebarIsCollapsed: PropTypes.bool,
> -               expandTheSidebar: PropTypes.func.isRequired,
> +               expandSidebar: PropTypes.func.isRequired,
>         };
>  
>         switchSites = ( event ) => {
>                 if ( isEnabled( 'nav-unification' ) && this.props.sidebarIsCollapsed ) {
> -                       this.props.expandTheSidebar();
> +                       this.props.expandSidebar();
>                 }
>                 event.preventDefault();
>                 event.stopPropagation();
> @@ -136,6 +136,6 @@ export default connect(
>         {
>                 recordGoogleEvent,
>                 setLayoutFocus,
> -               expandTheSidebar: () => expandSidebar(),
> +               expandSidebar,
>         }
>  )( localize( CurrentSite ) );
> ```
> 
> This could apply to collapse-sidebar.jsx as well. My goal is generally to remove one level of indirection and to make the mapdispatchtoprops function simpler. However I'm not tied to it 100% - maybe this is being done to make the two different concepts "CollapseSidebar" (the noun) and "collapseSidebar" (the verb) less confusing?

Well, while it works perfectly in `client/my-sites/current-site/index.jsx` it creates a naming collision in `client/my-sites/sidebar-unified/collapse-sidebar.jsx` 
![](https://cln.sh/lwmxro+)
where the noun - verb confusion is more prominent.

getdave commented on client/my-sites/current-site/index.jsx:
```diff
@@ -66,37 +72,54 @@ class CurrentSite extends Component {
-					</span>
-				) }
+				<div
+					role="button"
+					tabIndex={ 0 }
+					aria-hidden="true"
```
Can you explain if this is accessible and if so why? Often messing with tabindex directly is a cause of UX and a11y bugs.

cpapazoglou replied:
There is no need to be accessible, it is just an overlay above the 3 buttons area
![](https://cln.sh/dl6oyW+)
Tabbing through these 3 buttons (which have semantics) and clicking enter works.

getdave commented on client/my-sites/current-site/index.jsx:
```diff
@@ -66,37 +72,54 @@ class CurrentSite extends Component {
-				) }
+				<div
+					role="button"
+					tabIndex={ 0 }
```
```suggestion
					tabIndex="0"
```

cpapazoglou replied:
Fixed https://github.com/Automattic/wp-calypso/pull/46045/commits/4261e086b0f17a6b7a2114184e0a0aa625209c22

getdave commented on client/my-sites/sidebar-unified/collapse-sidebar.jsx:
```diff
@@ -0,0 +1,70 @@
+} ) => {
+	const collapsed = sidebarIsCollapsed && isWithinBreakpoint( '>800px' );
+
+	useEffect( () => {
```
As we're mutating the DOM here in a way which will change the appearance of the entire page we might need to consider `useLayoutEffect` as that will run before the browser has a chance to paint the DOM updates. Leaving as we have it will allow the browser to paint the component, then the effect will trigger and update the DOM changing the appearance.

cpapazoglou replied:
Great suggestion! Fixed https://github.com/Automattic/wp-calypso/pull/46045/commits/54126f462af6d30edc73d9161e4090cdd5b396bb

getdave commented on client/my-sites/sidebar-unified/collapse-sidebar.jsx:
```diff
@@ -0,0 +1,70 @@
+			className="collapse-sidebar__toggle"
+			onNavigate={ sidebarIsCollapsed ? expandTheSidebar : collapseTheSidebar }
+			label={ title }
+			link={ '' }
```
Do we need to pass the prop at all?

cpapazoglou replied:
Yeah, otherwise there is a console error `Failed prop type: The prop `link` is marked as required in `SidebarItem`, but its value is undefined` . 

Initially I used the `MySitesSidebarUnifiedMenu` which didn't need it but created other problems.

getdave commented on client/my-sites/sidebar-unified/collapse-sidebar.jsx:
```diff
@@ -0,0 +1,70 @@
+
+export default connect(
+	( state ) => {
+		return { sidebarIsCollapsed: getSidebarIsCollapsed( state ) };
```
You could use `useSelector` for this and avoid all the `connect` stuff. Is there a reason you use `connect`?

cpapazoglou replied:
Great suggestion too, I guess I only used it out of habit. Fixed https://github.com/Automattic/wp-calypso/pull/46045/commits/c0bc54eaac8bf9d7bb56d8bdb61130993bddcc29

getdave commented on client/state/ui/selectors/sidebar-visibility.js:
```diff
@@ -0,0 +1,3 @@
+export default function getSidebarIsCollapsed( state = false ) {
```
Why do we pass `false` as a default here? It would result in an error if the selector logic operated on it as `false` doesn't have a property `ui`...etc.

cpapazoglou replied:
Good catch, copy - paste error. Fixed https://github.com/Automattic/wp-calypso/pull/46045/commits/ae043b12094c73d282e32673cafa05d52f3eba5b

getdave commented on client/state/ui/test/reducer.js:
```diff
@@ -91,4 +100,30 @@ describe( 'reducer', () => {
 	} );
+
+	describe( '#getSidebarIsCollapsed()', () => {
+		test( 'should default to false', () => {
+			const state = sidebarIsCollapsed( undefined, {} );
+
+			expect( state ).to.be.false;
+		} );
```
I'd be tempted to create some initial state in all the tests and `deepFreeze` it so you know your reducers aren't mutating state.

cpapazoglou replied:
Done https://github.com/Automattic/wp-calypso/pull/46045/commits/86a3dc8c4f5dc220f516fbaaf897d04294641592

getdave commented on client/state/ui/test/reducer.js:
```diff
@@ -91,4 +100,30 @@ describe( 'reducer', () => {
+		} );
+
+		test( 'should be true when collapsed', () => {
+			const state = sidebarIsCollapsed( null, {
+				type: SIDEBAR_TOGGLE_VISIBILITY,
+				collapsed: true,
+			} );
+
+			expect( state ).to.be.true;
```
If we have some initial state here where `collapsed` is definitely set to `false` then we can be more confident the action passing through the reducer actually causes it to toggle to `true`.

cpapazoglou replied:
Done https://github.com/Automattic/wp-calypso/pull/46045/commits/86a3dc8c4f5dc220f516fbaaf897d04294641592

getdave reviewed (changes requested):
Thanks for picking this up again and driving it forward 👍 

I tested and found in general that this works very well. However, I noticed x2 things that were confusing to me (as someone coming to this with fresh eyes):

Firstly in Calypso if you have the sidebar collapsed and then you click on a Notice the sidebar is toggled open but you are also navigated to the Notice's content. For me this was disconcerting as I didn't know what the notice was about. I would have expected to be able to see the notice first and then click _again_ to trigger navigation.

![Screen Capture on 2020-10-21 at 15-23-32](https://user-images.githubusercontent.com/444434/96733485-69cb4380-13b1-11eb-932c-15cb55ecfc12.gif)

The second issue is in WPAdmin. When collapsed when I click the site switcher arrow button it navigates me to a 404 page. This isn't what I expected.

![Screen Capture on 2020-10-21 at 15-04-32 (1)](https://user-images.githubusercontent.com/444434/96733842-ce869e00-13b1-11eb-9627-a40ccf7f6807.gif)

Also noticed some code issues which we can discuss. This PR is becoming quite large and I feel this is really close. Therefore I might even say if we can address the _code_ issues, we can merge and then defer the other points to followups Issues. 

Thanks again 🥇

cpapazoglou commented:
Thanks @getdave for you thorough review!

> I would have expected to be able to see the notice first and then click again to trigger navigation.
I totally understand your point! Let's iterate on it.

> The second issue is in WPAdmin. When collapsed when I click the site switcher arrow button it navigates me to a 404 page. This isn't what I expected.

Very good catch, I don't what I had in mind and why I didn't test it.. Updated the patch D50420-code

cpapazoglou commented:
Rebased with master

getdave commented:
@cpapazoglou Thanks for making those changes. I haven't had a chance to re-manual test this, but I will do so first thing tomorrow unless someone gets there first.

🥇

cpapazoglou commented:
Deploed r215609-wpcom
